### ISSUE TITLE
MenuItem::activate stalls (complete) system rendering

### ISSUE THREAD

**Issue by antis81** [slint-ui/slint#9305]
### Bug Description

I have a desktop application with the infamous "File" menu and the requirement to show the open/save dialog (`rfd` crate). The code was written before MenuBar was available (Slint < 1.12) and I worked around this putting buttons in a side bar.

A minimal backend should trigger this bug and stall the desktop envionment (requires rfd crate).

```rust
slint::include_modules!();

fn main() {
  let ui = App::new().unwrap();
  
  ui.window.on_open_file({
    move || {
      let selected_file = rfd::FileDialog::new()
        .set_title("Select a hm64 file")
        .add_filter("hm64 files", &["hm64"])
        .pick_file();
  });

  ui.run().unwrap();
}
```

### Reproducible Code (if applicable)

```slint
import { AboutSlint, HorizontalBox, StyleMetrics, Button, VerticalBox } from "std-widgets.slint";
export component App inherits Window {
    // backend interface
    callback open-file();
    // callback save-file();  // excluded for the example


    // new implementation (Slint >= 1.12)
    MenuBar {
        Menu {
            title: "File";
            MenuItem {
                title: "Open file\n(stalls desktop rendering!)";
                activated => {
                    root.open-file();
                }
            }
        }
    }

    layout := VerticalBox {
        AboutSlint {
            max-height: 150px;
        }

        HorizontalBox {
            // left side bar
            VerticalLayout {
                spacing: StyleMetrics.layout-spacing;
                // old implementation (slint < 1.12)
                Button {
                    text: "Open file\n(works just fine)";
                    clicked => {
                        root.open-file();
                    }
                }
            }

            // content
            Rectangle {
                horizontal-stretch: 1;
                background: lightcyan;

                Text {
                    horizontal-alignment: center;
                    vertical-alignment: center;
                    color: gray;
                    text: "Content…";
                }
            }
        }
    }
}
```

### Environment Details

- Slint Version: 1.12.1
- Platform/OS: Linux (LXQt)
- Programming Language: Rust
- Backend/Renderer: Qt, X11 (did not check Wayland)

### Product Impact

Product stability/reliability

**Comment by ogoffart** [comment]
Thanks for filing a bug.

I haven't tried this code, but i suspect the following is happening:

For the popup menu, the Qt backend is opening a Popup Window. Popup Window grab the mouse and keyboard for the entire system until it is closed.
In the activate caallback, we call rfd which probably blocks the main thread, and so the popup never gets closed.

I think the work around for this would be to use a thread or some async API from rfd in combination with slint::spawn_local.

Not sure if we can actually fix this in the Qt backend or how to fix it.

**Comment by antis81** [comment]
Okay think I understand the problem there. Wouldn't the async/thread workaround apply to pretty much any action called from within a menu?

In my use case the action itself is (always) async anyway (no return value and the data is completely managed by the backend). Also I can use the `rfd` async API to further "de-couple" this…

While writing these lines an idea popped in my mind:
Do I understand this right that callbacks are actual callbacks (direct function calls)? Then Slint could maybe have "actions" (or simply events). Syntactically I find this more readable than "async callback abc()". Anyway those would always go to the event loop and require async (better: de-coupled) handling (to keep the event loop "fluffy"). In the Qt world this translates to `Qt::QueuedConnection` vs `Qt::DirectConnection`. (This would also solve "your headache" thinking too much about the item tree… 😄). Do you think this could be an interesting feature?
